lib/src/files.rs

```diff
@@ -207,11 +207,8 @@ struct SyncRegion {
     right: Range<usize>,
 }
 
-// TODO: Should we require `add.len() == removes.len() + 1`? If that condition
-// is false, it effectively means that we should pretend that there are empty
-// strings in `removes` or `adds` to make it true. Maybe we should have to
-// caller make it explicitly that way.
```

Comment:
> If that condition // is false, it effectively means that we should pretend that there are empty // strings in `removes` or `adds` to make it true.

I don't really understand what empty strings mean in this context. I'll read this function more closely at a later time.

Reply:
I think I meant that this function (before this commit) would merge any combination of adds and removes as if those arrays had been padded with empty strings. For example `merge(&[], &["a", "b"])` would be equivalent to `merge(&[""], &["a", "b"])` and `merge(&["a"], &["b"])` would be equivalent to `merge(&["a"], &["b", ""])`. In practice, that means that it behaves as if the file contents had been empty in the "missing" terms.

Reply:
> that means that it behaves as if the file contents had been empty in the "missing" terms.

Ah, thanks. I'll reread with this interpretation.

Reply:
> For example merge(&[], &["a", "b"]) would be equivalent to merge(&[""], &["a", "b"]) and merge(&["a"], &["b"]) would be equivalent to merge(&["a"], &["b", ""])

Hm, does this happen in `Diff::for_tokenizer(&diff_inputs, &diff::find_line_ranges)`? In there, I see some "padding"

```
        unchanged_regions.push(UnchangedRange {
            base_range: base_input.len()..base_input.len(),
            offsets,
        });
```

but it also seems to assume that there is exactly one remove:

```
        assert!(!inputs.is_empty());
        let base_input = inputs[0];
        let other_inputs = inputs.iter().skip(1).copied().collect_vec();
        // First tokenize each input
        let base_token_ranges: Vec<Range<usize>> = tokenizer(base_input);
        let other_token_ranges: Vec<Vec<Range<usize>>> = other_inputs
            .iter()
            .map(|other_input| tokenizer(other_input))
            .collect_vec();
```

Reply:
Hm, looks like it doesn't actually assume one remove. I tried a bunch of test cases with multiple removes and couldn't get it to misbehave. I guess it works because `Diff::for_tokenizer` just calculates the hunks by comparing everything against an arbitrary base, but it's this function's responsibility to make sure the hunks are represented correctly as adds/removes.

Reply:
>  This block seems to be what adds the empty strings (or byte vectors)

Hm, I thought that "pretend the file was empty during a conflict" was only about making conflicts easier to grok (in response to a complaint of mine about moved files). At any rate, don't feel the need to look this up on my behalf :)